spec/abstract/KHeap_A.thy

```diff
@@ -304,50 +347,11 @@ definition
 where
   "in_release_queue tcb_ptr \<equiv> \<lambda>s. tcb_ptr \<in> set (release_queue s)"
 
-definition
-  is_schedulable :: "obj_ref \<Rightarrow> ('z::state_ext state, bool) nondet_monad"
-where
-  "is_schedulable tcb_ptr \<equiv> do
-    tcb \<leftarrow> gets_the $ get_tcb tcb_ptr;
-    if Option.is_none (tcb_sched_context tcb)
-    then return False
-    else do
-      sc \<leftarrow> get_sched_context $ the $ tcb_sched_context tcb;
-      is_runnable \<leftarrow> return (runnable (tcb_state tcb));
-      in_release_q \<leftarrow> gets $ in_release_queue tcb_ptr;
-      return (is_runnable \<and> sc_active sc \<and> \<not>in_release_q)
-    od
-  od"
-
-definition
-  is_sc_active :: "obj_ref \<Rightarrow> 'z::state_ext state \<Rightarrow> bool"
-where
-  "is_sc_active sp \<equiv> (\<lambda>s.
-     case kheap s sp of
-       Some (SchedContext sc _) \<Rightarrow> sc_active sc
-     | _ \<Rightarrow> False)"
-
-definition
-  is_schedulable_opt :: "obj_ref \<Rightarrow> 'z::state_ext state \<Rightarrow> bool option"
-where
-  "is_schedulable_opt tcb_ptr \<equiv> \<lambda>s.
-     case get_tcb tcb_ptr s of None \<Rightarrow> None
-     | Some tcb \<Rightarrow>
-       (case tcb_sched_context tcb of None => Some False
-        | Some sc_ptr =>
-            Some (runnable (tcb_state tcb) \<and> (is_sc_active sc_ptr s)
-            \<and> \<not>(in_release_queue tcb_ptr s)))"
-
-definition
-  schedulable :: "obj_ref \<Rightarrow> 'z::state_ext state \<Rightarrow> bool"
-where
-  "schedulable tcb_ptr \<equiv> \<lambda>s.
-     case get_tcb tcb_ptr s of None \<Rightarrow> False
-     | Some tcb \<Rightarrow>
-       (case tcb_sched_context tcb of None => False
-        | Some sc_ptr =>
-            (runnable (tcb_state tcb) \<and> (is_sc_active sc_ptr s)
-              \<and> \<not>(in_release_queue tcb_ptr s)))"
+definition schedulable :: "obj_ref \<Rightarrow> 'z::state_ext state \<Rightarrow> bool" where
+  "schedulable t s \<equiv>
+   (runnable |< (tcbs_of s ||> tcb_state)) t
+   \<and> (sc_active |< (tcbs_of s |> tcb_sched_context |> scs_of s)) t
+   \<and> t \<notin> set (release_queue s)"
```

Comment:
This is so much nicer than the triplicate terrible manual case distinctions from before. Thanks for taking on that cleanup task.